packages/openapi-v3/spike.md

```diff
@@ -0,0 +1,71 @@
+## Improve the UX of @requestBody()
+
+The original discussion is tracked in issue [Spike: simplify requestBody annotation with schema options](https://github.com/strongloop/loopback-next/issues/2654).
+
+The current @requestBody() can only
+
+- takes in an entire request body specification with very nested media type objects
+or
+- generate the schema inferred from the parameter type
+
+To simplify the signature, this spike PR introduces a 2nd parameter `schemaOptions` to configure the schema. The new decorator `newRequestBody1` is written in file 'request-body.options1.decorator.ts'
+
+### Create - exclude properties
+
+Take "Create a new product with excluded properties" as an example:
+
+```ts
+// Provide the description as before
+const requestBodySpec = {
+  description: 'Create a product',
+  required: true,
+};
+
+// Provide the options that configure your schema
+const excludeOptions = {
+  // Using advanced ts types like `Exclude<>`, `Partial<>` results in
+  // `MetadataInspector.getDesignTypeForMethod(target, member)` only
+  // returns `Object` as the param type, which loses the model type's info.
+  // Therefore user must provide the model type in options.
+  [TS_TYPE_KEY]: Product,
+  // Make sure you give the custom schema a unique schema name,
+  // this name will be used as the reference name
+  // like `$ref: '#components/schemas/ProductWithoutID'`
+  schemaName: 'ProductWithoutID',
+  // The excluded properties
+  exclude: ['id']
+}
```

Comment:
I am quite unhappy about this design.

It's seems to be based on the old `ts-type` approach that we are slowly moving away from (see e.g. https://github.com/strongloop/loopback-next/pull/3402) and using `getModelSchemaRef` instead.

Let's focus on a solution that would leverage `getModelSchemaRef` and allow developers using different ORM frameworks like TypeORM to use their own helper building model-schemas from their model definitions.

Reply:
@bajtos It's ok, I was about to switch to your proposal which uses `getModelSchemaRef` to generate the OpenAPI schema
```
class MyController1 {
  @post('/Product')
  create(@newRequestBody1(
    requestBodySpec,
    getModelSchemaRef(Product, {exclude: ['id']}),
  ) product: Exclude<Product, ['id']>) { }
}
```

but now I am wondering, what is the main purpose of SIMPLIFYING the decorator? To "avoid using very nested object to provide the declarative schema config(which is specific to loopback 4 core) - option1" or to "leverage the schema builder to generate the OAI3 schema that any ORM can consume - option2"?

My current design serves option1, but if our purpose is option2, I will try @bajtos's proposal.

cc @strongloop/loopback-maintainers any thought/preference?

Reply:
What's the minimum way to have a valid `requestBody` spec? I assume the new decorator is designed to allow so:

- Use a TS class such as `Product` to build the schema
- Use options such as `exclude` or `optional` to tweak the schema

I would expect something like the following to represent `Exclude<Product, ['id']>`:
```ts
class MyController1 {
  @post('/Product')
  create(@requestBody.fromType(Product, {exclude: ['id']}) product: Exclude<Product, ['id']>) { }
}
```
```

Reply:
@raymondfeng see [the spec definition](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#request-body-object)

At least we need two more fields:
- `description`: currently requires user to provide in `@requestBody`, but in the future it can be read from model's metadata
- `required`: a boolean value

And the nested `content` object contains more fields like `examples` besides `schema`, see comment https://github.com/strongloop/loopback-next/pull/3398#issuecomment-513726039, the discussion of the `content` spec is at the bottom.

Reply:
> what is the main purpose of SIMPLIFYING the decorator?

I created that user story with the intention to find a more ergonomic solution for describing request bodies. IMO, controllers methods using the default body parser should not need to repeat the fact that the request body accepts content of type `application/json`.

See e.g. what Raymond proposed in his comment:

```ts
@requestBody.fromType(Product, {exclude: ['id']})
```

and what I proposed in elsewhere in this pull request:

```ts
@requestBody(MyModel, {
  // parameter-definition options
  description: 'description of the request body parameter',
  required: true,
  // schema options
  partial: true,
  optional: ['categoryId'],
  exclude: ['id'],
})
```

The examples show a reasonably complex case. It may be not obvious that they are also making the default use case super simple - and that was the original goal!

```ts
@requestBody.fromType(Product);
// or
@requestBody(Product)
```

---

In the future, we may add support for other content-types, e.g. XML or [JSON API](https://jsonapi.org/format/#content-negotiation). It would be nice if application developers didn't need to edit all their `@requestBody` calls to add those new content types when they become supported. This isn't required, just something to be aware of.